Remove native_functions.yaml dependency from TensorModeKernel.cu

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67875
* #67874
* #67721
* #66980
* #66979
* #67216
* #66914
* __->__ #66913

Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

peterbell10 committed:
Remove native_functions.yaml dependency from TensorModeKernel.cu

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

dagitses commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -13,18 +14,27 @@
 namespace at {
 namespace native {
 
+static inline int64_t ensure_nonempty_dim(int64_t dim) {
```
rather than duplicating these, should you extract them from ReduceOpsUtils.h into a new header that only depends on TensorBase? ReduceOpsUtils.h ought to be able to depend on that to reduce churn, no?

dagitses commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -69,11 +83,11 @@ void calculate_mode(
 
-  auto keys_ptr = thrust::device_pointer_cast(keys.data_ptr<scalar_t>());
-  auto counts_ptr = thrust::device_pointer_cast(counts.data_ptr<int64_t>());
+  auto keys_ptr = thrust::device_pointer_cast(static_cast<scalar_t*>(keys.get()));
```
there's a common pattern here used at least three times: cuda allocate N elements of type T. This should probably be a utility to avoid the duplicating of the type and the potential to get that wrong.

The wrench is that the storage's lifetime is managed by result from the allocator.

What roughly makes sense to me:
```
template <typename T>
class CudaArray {
 public:
  CudaArray(Allocator& allocator, std::size_t size) : CudaArray(allocator.allocate(size * sizeof(T))) {}
 
  T* as_ptr() { return ptr_; }
  // ... blah blah
 private:
  explicit CudaAllocator(DataPtr data) : data_(std::move(data)), ptr_(thrust::device_pointer_cast(static_cast<T*>(data_.get()))) {}
  DataPtr data_;
  T* ptr_;
};

dagitses replied:
thanks, i like the way you did it better.

Commit:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

peterbell10 commented on c10/core/DeviceArray.h:
```diff
@@ -0,0 +1,21 @@
+  DeviceArray(c10::Allocator& allocator, size_t size)
+      : data_ptr_(allocator.allocate(size * sizeof(T))) {
+    static_assert(std::is_pod<T>::value, "T must be 'plain old data'");
+  }
```
>check alignment so that get() is ok below?

I don't think that can be done statically as we don't know the allocator's alignment guarantees. (CUDA provides higher alignment by default).

>do we need to construct and destroy the elements with placement new even if they are POD?

I think POD types qualify as [implicit-lifetime types](https://en.cppreference.com/w/cpp/language/lifetime#Implicit-lifetime_types) which means its lifetime begins at allocation even if no constructor is ever called. Also, object lifetime unconditionally ends at deallocation so calling the destructor is only required if it's non-trivial but `is_pod` guarantees it is trivial.

dagitses replied:
> > check alignment so that get() is ok below?
> 
> I don't think that can be done statically as we don't know the allocator's alignment guarantees. (CUDA provides higher alignment by default).
I mean does it make sense to `assert()` that the alignment is suitable. 
> 
> > do we need to construct and destroy the elements with placement new even if they are POD?
> 
> I think POD types qualify as [implicit-lifetime types](https://en.cppreference.com/w/cpp/language/lifetime#Implicit-lifetime_types) which means its lifetime begins at allocation even if no constructor is ever called. Also, object lifetime unconditionally ends at deallocation so calling the destructor is only required if it's non-trivial but `is_pod` guarantees it is trivial.
Isn't this guarantee only true for specific allocation functions? Is the CUDA allocator included in this list?

Can the requirement of this be loosened to `std::is_trivial`?

peterbell10 replied:
>Isn't this guarantee only true for specific allocation functions? 

You can read the [cppreference text here](https://en.cppreference.com/w/cpp/language/object#Object_creation). The allocations functions listed are I believe every standard way to allocate raw memory (`operator new`, `malloc` and arrays of bytes). I believe the intention is to support C-style code where constructors don't exist.

> Is the CUDA allocator included in this list?

CUDA isn't really relevant to the C++ standard here because the C++ compiler can't dereference CUDA memory at all. The pattern of just assigning to CUDA memory is used ubiquitously in CUDA examples so is clearly supported.

> Can the requirement of this be loosened to `std::is_trivial`?

I think that's right. Changed and rebased on `viable/strict`.

dagitses replied:
> > > check alignment so that get() is ok below?
> > 
> > 
> > I don't think that can be done statically as we don't know the allocator's alignment guarantees. (CUDA provides higher alignment by default).
> 
> I mean does it make sense to `assert()` that the alignment is suitable.

I messed up the quoted reply earlier, what do you think of this?
> 
> > > do we need to construct and destroy the elements with placement new even if they are POD?
> > 
> > 
> > I think POD types qualify as [implicit-lifetime types](https://en.cppreference.com/w/cpp/language/lifetime#Implicit-lifetime_types) which means its lifetime begins at allocation even if no constructor is ever called. Also, object lifetime unconditionally ends at deallocation so calling the destructor is only required if it's non-trivial but `is_pod` guarantees it is trivial.
> 
> Isn't this guarantee only true for specific allocation functions? Is the CUDA allocator included in this list?
> 
> Can the requirement of this be loosened to `std::is_trivial`?

dagitses replied:
> > Isn't this guarantee only true for specific allocation functions?
> 
> You can read the [cppreference text here](https://en.cppreference.com/w/cpp/language/object#Object_creation). The allocations functions listed are I believe every standard way to allocate raw memory (`operator new`, `malloc` and arrays of bytes). I believe the intention is to support C-style code where constructors don't exist.
> 
> > Is the CUDA allocator included in this list?
> 
> CUDA isn't really relevant to the C++ standard here because the C++ compiler can't dereference CUDA memory at all. The pattern of just assigning to CUDA memory is used ubiquitously in CUDA examples so is clearly supported.

I'm not super worried about this, but the cppreference page seems to indicate that the implicit lifetimes are limited to the enumerated allocator functions. Is `c10::Allocator` guaranteed to use one of them?
> 
> > Can the requirement of this be loosened to `std::is_trivial`?
> 
> I think that's right. Changed and rebased on `viable/strict`.

peterbell10 replied:
>I mean does it make sense to `assert()` that the alignment is suitable.
Okay, I've added an assert.

> I'm not super worried about this, but the cppreference page seems to indicate that the implicit lifetimes are limited to the enumerated allocator functions. Is c10::Allocator guaranteed to use one of them?

The CPU allocator uses `posix_memalign` which isn't part of standard C++ so can't be included in that list. However, it's notionally the same as `malloc` so I don't think it will be an issue.

dagitses replied:
sounds good, thanks!

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

dagitses commented on aten/src/ATen/native/NonEmptyUtils.h:
```diff
@@ -0,0 +1,27 @@
+
+namespace at { namespace native {
+
+static inline int64_t ensure_nonempty_dim(int64_t dim) {
```
remove static from all of these?

dagitses commented on aten/src/ATen/native/NonEmptyUtils.h:
```diff
@@ -0,0 +1,27 @@
+
+using IdxVec = std::vector<int64_t>;
+static inline IdxVec ensure_nonempty_vec(IdxVec vec) {
+  if (vec.size() == 0) {
```
vec.empty()

dagitses commented on aten/src/ATen/native/cuda/TensorModeKernel.cpp:
```diff
@@ -0,0 +1,101 @@
+
+  // Resize output value, index Tensors to appropriate sizes (i.e. the same as
+  // the input Tensor, except at dim=dimension, the size is 1)
+  self_sizes[dim] = 1;
```
`assert(dim < self_sizes.size());`

dagitses replied:
It appears that `TORCH_INTERNAL_ASSERT_DEBUG_ONLY` does not suppress the linter, but `assert` does.

dagitses replied:
Also this flags a sign-compare problem now, of course. so after the comparison to 0 add `static_cast<std::size_t>(dim)` in the comparison against size.

peterbell10 replied:
Alright, changed to `assert(0 <= dim && static_cast<size_t>(dim) < self_sizes.size())`

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).

Commit:
Update on "Remove native_functions.yaml dependency from TensorModeKernel.cu"


Differential Revision: [D31856102](https://our.internmc.facebook.com/intern/diff/D31856102)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856102).
